src/fairseq2/recipes/wav2vec2/train.py

```diff
@@ -0,0 +1,360 @@
+
+    fsdp_wrap_granularity: Literal["layer", "stack", "model"] = "stack"
+    """The granularity at which to wrap the model."""
+
+    torch_compile: bool = False
+    """If ``True``, applies ``torch.compile()`` to the encoder. (experimental)"""
+
+    # Optimizer, LR, and Loss
+    lr: float = 5e-04
+    """The initial (post-warm-up) learning rate."""
+
+    warmup_steps: int = 32_000
+    """The number of warmup steps for polynomial LR scheduler."""
+
+    betas: Tuple[float, float] = (0.9, 0.98)
+    """The coefficients of AdamW."""
+
+    eps: float = 1e-06
+    """Adam's epsilon parameter."""
+
+    weight_decay: float = 0.01
+    """The weight decay parameter used for regularization."""
+
+    max_gradient_norm: Optional[float] = None
+    """The maximum gradient norm. If ``None``, no clipping will be applied."""
+
+    fp16_loss_scale: Tuple[float, float] = (128.0, 0.0001)
+    """The initial and minimum loss scale for fp16 training."""
+
+    gradient_accumulation: int = 1
+    """The number of steps to accumulate gradients before an optimizer update."""
+
+    # Regime
+    max_num_steps: int = 400_000
+    """The maximum number of steps to train for."""
+
+    max_num_data_epochs: Optional[int] = None
+    """The maximum number of data epochs to train for."""
+
+    validate_every_n_steps: int = 1000
+    """The step interval at which to validate the model."""
+
+    checkpoint_after_n_steps: int = 0
+    """The number of steps after which to start checkpointing."""
+
+    checkpoint_every_n_steps: int = 25_000
+    """The step interval at which to checkpoint."""
+
+    publish_metrics_every_n_steps: int = 1000
+    """The step interval at which to publish metrics."""
+
+    # Misc
+    seed: int = 2
+    """The random number generator seed to use."""
+
+    profile: Optional[Tuple[int, int]] = None
+    """The number of steps that the PyTorch profiler should skip and then record."""
+
+    monitored_gang: bool = False
+    """If ``True``, puts a monitored barrier before every collective call."""
+
+    anomaly_detection: bool = False
+    """If ``True``, enables the anomaly detection feature of ``torch.autograd``."""
+
+
+wav2vec2_train_presets = ConfigRegistry[Wav2Vec2TrainConfig]()
+
+wav2vec2_train_preset = wav2vec2_train_presets.decorator
+
+
+@wav2vec2_train_preset("base_960h")
+def _base_960h() -> Wav2Vec2TrainConfig:
+    return Wav2Vec2TrainConfig()
+
+
+def load_wav2vec2_trainer(
+    config: Wav2Vec2TrainConfig, output_dir: Path
+) -> Trainer[Tensor]:
+    """Load a :class:`Trainer` for wav2vec 2.0 model training."""
+    wall_watch = Stopwatch(start=True)
+
+    gang = setup_root_gang(log, monitored=config.monitored_gang)
+
+    checkpoint_manager = FileCheckpointManager(output_dir.joinpath("checkpoints"), gang)
+
+    # Load the dataset.
+    dataset_card = retrieve_asset_card(config.dataset)
+
+    log.info("Loading {} speech dataset.", dataset_card.name)
+
+    dataset = load_speech_dataset(config.dataset)
+
```

Comment:
and also here? I think the logging can be line-attached to the function that it suprvise